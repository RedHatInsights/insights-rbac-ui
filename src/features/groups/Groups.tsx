import React, { Fragment, Suspense, useContext, useMemo, useState } from 'react';
import { Outlet } from 'react-router-dom';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';
import { useIntl } from 'react-intl';

// PatternFly imports
import { DataView, DataViewState } from '@patternfly/react-data-view';
import { DataViewToolbar } from '@patternfly/react-data-view/dist/dynamic/DataViewToolbar';
import DataViewFilters from '@patternfly/react-data-view/dist/cjs/DataViewFilters';
import { DataViewTextFilter } from '@patternfly/react-data-view';
import { Button } from '@patternfly/react-core/dist/dynamic/components/Button';
import { Stack, StackItem } from '@patternfly/react-core/dist/dynamic/layouts/Stack';
import { Pagination } from '@patternfly/react-core/dist/dynamic/components/Pagination';
import { BulkSelect } from '@patternfly/react-component-groups/dist/dynamic/BulkSelect';
import { Dropdown } from '@patternfly/react-core/dist/dynamic/components/Dropdown';
import { DropdownList } from '@patternfly/react-core/dist/dynamic/components/Dropdown';
import { DropdownItem } from '@patternfly/react-core/dist/dynamic/components/Dropdown';
import { MenuToggle } from '@patternfly/react-core/dist/dynamic/components/MenuToggle';
import EllipsisVIcon from '@patternfly/react-icons/dist/js/icons/ellipsis-v-icon';

// Red Hat components
import Section from '@redhat-cloud-services/frontend-components/Section';
import { PageLayout, PageTitle } from '../../components/layout/PageLayout';
import { AppLink } from '../../components/navigation/AppLink';
import { useChrome } from '@redhat-cloud-services/frontend-components/useChrome';

// Redux imports
import { fetchGroups } from '../../redux/groups/actions';

// Helper imports
import { defaultAdminSettings, defaultSettings } from '../../helpers/pagination';
import { getBackRoute } from '../../helpers/navigation';

// Context and types
import PermissionsContext from '../../utilities/permissionsContext';
import { PER_PAGE_OPTIONS } from '../../helpers/pagination';
import messages from '../../Messages';
import pathnames from '../../utilities/pathnames';
import type { Group, RBACStore } from './types';
import { GroupsEmptyState } from './components/GroupsEmptyState';
import { GroupsTableContent } from './components/GroupsTableContent';
import { GroupsSkeletonTable } from './components/GroupsSkeletonTable';
// Custom hooks
import { useGroupsHandlers } from './useGroupsHandlers';
import { useUrlPaginationSync } from './useUrlPaginationSync';
import { useUrlFiltersSync } from './useUrlFiltersSync';
import { useInitialGroupsDataLoad } from './useInitialGroupsDataLoad';
import { useRouteStateManagement } from './useRouteStateManagement';
import { useTableColumns } from './useTableColumns';
import './Groups.scss';

// Types
interface GroupsProps {}

export const Groups: React.FC<GroupsProps> = () => {
  const intl = useIntl();
  const dispatch = useDispatch();
  const chrome = useChrome();
  const fetchData = (options: { limit?: number; offset?: number; filters?: Record<string, unknown>; sortBy?: string; orderBy?: string }) => {
    // Convert orderBy string to enum if needed
    const { orderBy, ...restOptions } = options;
    const fetchOptions = {
      ...restOptions,
      usesMetaInURL: true,
      chrome,
      platformDefault: false,
      adminDefault: false,
      ...(orderBy && { orderBy: orderBy as any }), // Cast to any since @redhat-cloud-services/rbac-client has broken enum types
    };
    return dispatch(fetchGroups(fetchOptions));
  };

  // User permission context for admin/user access control
  const { orgAdmin, userAccessAdministrator } = useContext(PermissionsContext);
  const isAdmin = orgAdmin || userAccessAdministrator;

  // Table column definitions
  const columns = useTableColumns();

  // Component state for table sorting and row selection
  const [sortByState, setSortByState] = useState({ index: Number(isAdmin), direction: 'asc' as 'asc' | 'desc' });
  const [selectedRows, setSelectedRows] = useState<Group[]>([]);
  const [removeGroupsList, setRemoveGroupsList] = useState<Group[]>([]);
  const [expanded, setExpanded] = useState<Record<string, number | boolean>>({});
  const [isBulkActionsOpen, setIsBulkActionsOpen] = useState(false);

  const orderBy = `${sortByState?.direction === 'desc' ? '-' : ''}${columns[sortByState?.index - Number(isAdmin)]?.key || 'name'}`;

  // Redux state selectors for groups data and loading state
  const { groups, pagination, filters, isLoading } = useSelector(
    ({
      groupReducer: {
        groups: { data, filters, pagination },
        isLoading,
        adminGroup,
        systemGroup,
      },
    }: RBACStore) => ({
      groups: [
        ...(adminGroup?.name?.match(new RegExp(filters?.name || '', 'i')) ? [adminGroup] : []),
        ...(systemGroup?.name?.match(new RegExp(filters?.name || '', 'i')) ? [systemGroup] : []),
        ...(data?.filter(
          ({ platform_default, admin_default }: { platform_default?: boolean; admin_default?: boolean } = {}) => !(platform_default || admin_default),
        ) || []),
      ],
      pagination: {
        limit: pagination?.limit ?? (orgAdmin ? defaultAdminSettings : defaultSettings).limit,
        offset: pagination?.offset ?? (orgAdmin ? defaultAdminSettings : defaultSettings).offset,
        count: pagination?.count,
        redirected: pagination?.redirected,
      },
      filters: filters,
      isLoading,
    }),
    shallowEqual,
  );

  const [filterValue, setFilterValue] = useState(filters?.name || '');

  // URL and route state management hooks
  useUrlPaginationSync(pagination.limit, pagination.offset, pagination.count, pagination.redirected);
  useUrlFiltersSync(filterValue);
  useInitialGroupsDataLoad({
    pagination,
    filterValue,
    onFilterValueChange: setFilterValue,
    fetchData,
  });
  useRouteStateManagement(pagination, filterValue);

  // Data transformation: merge admin/system groups for admin users, standard groups for others
  const data = useMemo(
    () =>
      groups.map((group: Group) =>
        group.platform_default || group.admin_default ? { ...group, principalCount: `All${group.admin_default ? ' org admins' : ''}` } : group,
      ),
    [groups],
  );

  // Computed values that handlers need
  const selectableItemsCount = data.filter((item) => !(item.platform_default || item.admin_default)).length;
  const currentPage = Math.floor(pagination.offset / pagination.limit) + 1;
  const totalCount = pagination.count || data.length;
  const hasActiveFilters = filterValue !== '';

  // Use handlers hook
  const {
    onExpand,
    handleSort,
    handleBulkSelect,
    handleEdit,
    handleDelete,
    handlePageChange,
    handlePerPageChange,
    handleFiltersChange,
    handleClearAllFilters,
  } = useGroupsHandlers({
    isAdmin,
    columns,
    expanded,
    onExpandedChange: setExpanded,
    onSelectedRowsChange: setSelectedRows,
    onRemoveGroupsChange: setRemoveGroupsList,
    onSortChange: setSortByState,
    onFilterValueChange: setFilterValue,
    data,
    pagination,
    orderBy,
    totalCount,
    filterValue,
    fetchData,
  });

  // Bulk deletion state: determines if all rows are being removed to manage filter reset behavior
  const removingAllRows = pagination.count === removeGroupsList.length;

  const activeState = isLoading ? DataViewState.loading : data.length === 0 ? DataViewState.empty : undefined;

  return (
    <Fragment>
      <Stack className="rbac-c-groups">
        <StackItem>
          <PageLayout>
            <PageTitle title={intl.formatMessage(messages.groups)} />
          </PageLayout>
        </StackItem>
        <StackItem>
          <Section type="content" id="tab-groups">
            <DataView activeState={activeState}>
              <DataViewToolbar
                bulkSelect={
                  isAdmin && data.length > 0 ? (
                    <BulkSelect selectedCount={selectedRows.length} totalCount={selectableItemsCount} onSelect={handleBulkSelect} />
                  ) : undefined
                }
                actions={
                  isAdmin && selectedRows.length > 0 ? (
                    <Dropdown
                      isOpen={isBulkActionsOpen}
                      onOpenChange={setIsBulkActionsOpen}
                      toggle={(toggleRef) => (
                        <MenuToggle
                          ref={toggleRef}
                          aria-label="bulk actions toggle"
                          variant="plain"
                          onClick={() => setIsBulkActionsOpen(!isBulkActionsOpen)}
                        >
                          <EllipsisVIcon />
                        </MenuToggle>
                      )}
                    >
                      <DropdownList>
                        <DropdownItem
                          onClick={() => {
                            if (selectedRows.length === 1) {
                              handleEdit(selectedRows[0].uuid);
                            }
                            setIsBulkActionsOpen(false);
                          }}
                          isDisabled={selectedRows.length !== 1}
                        >
                          {intl.formatMessage(messages.edit)}
                        </DropdownItem>
                        <DropdownItem
                          onClick={() => {
                            handleDelete(selectedRows);
                            setIsBulkActionsOpen(false);
                          }}
                          isDisabled={selectedRows.length === 0}
                        >
                          {intl.formatMessage(messages.delete)}
                        </DropdownItem>
                      </DropdownList>
                    </Dropdown>
                  ) : isAdmin ? (
                    <AppLink to={pathnames['add-group'].link}>
                      <Button ouiaId="create-group-button" variant="primary">
                        {intl.formatMessage(messages.createGroup)}
                      </Button>
                    </AppLink>
                  ) : undefined
                }
                pagination={
                  totalCount > 0 ? (
                    <Pagination
                      perPageOptions={PER_PAGE_OPTIONS}
                      itemCount={totalCount}
                      page={currentPage}
                      perPage={pagination.limit}
                      onSetPage={(_event, page) => handlePageChange(page)}
                      onPerPageSelect={(_event, perPage) => handlePerPageChange(perPage)}
                    />
                  ) : undefined
                }
                filters={
                  <DataViewFilters onChange={(_event, values) => handleFiltersChange(values)} values={{ name: filterValue }}>
                    <DataViewTextFilter
                      filterId="name"
                      title={intl.formatMessage(messages.name)}
                      placeholder={`Filter by ${intl.formatMessage(messages.name).toLowerCase()}`}
                    />
                  </DataViewFilters>
                }
                clearAllFilters={handleClearAllFilters}
              />

              {/* Manual state handling for compound expansion */}
              {isLoading ? (
                <GroupsSkeletonTable isAdmin={isAdmin} rowsCount={10} />
              ) : data.length === 0 ? (
                <GroupsEmptyState
                  hasActiveFilters={hasActiveFilters}
                  titleText={`Configure ${intl.formatMessage(messages.groups).toLowerCase()}`}
                  isAdmin={isAdmin}
                />
              ) : (
                <GroupsTableContent
                  data={data}
                  isAdmin={isAdmin}
                  sortByState={sortByState}
                  selectedRows={selectedRows}
                  selectableItemsCount={selectableItemsCount}
                  expanded={expanded}
                  onExpandedChange={setExpanded}
                  onSelectedRowsChange={setSelectedRows}
                  onRemoveGroupsChange={setRemoveGroupsList}
                  handleBulkSelect={handleBulkSelect}
                  handleSort={handleSort}
                  onExpand={onExpand}
                />
              )}

              <DataViewToolbar
                pagination={
                  totalCount > 0 ? (
                    <Pagination
                      perPageOptions={PER_PAGE_OPTIONS}
                      itemCount={totalCount}
                      page={currentPage}
                      perPage={pagination.limit}
                      onSetPage={(_event, page) => handlePageChange(page)}
                      onPerPageSelect={(_event, perPage) => handlePerPageChange(perPage)}
                    />
                  ) : undefined
                }
              />
            </DataView>

            <Suspense fallback={<div>Loading...</div>}>
              <Outlet
                context={{
                  pagination,
                  filters,
                  [pathnames['add-group'].path]: {
                    orderBy,
                    postMethod: (config: Record<string, unknown>) => {
                      setFilterValue('');
                      fetchData(config);
                    },
                  },
                  [pathnames['edit-group'].path]: {
                    postMethod: (config: Record<string, unknown>) => {
                      setFilterValue('');
                      fetchData({ ...config, orderBy });
                    },
                    cancelRoute: getBackRoute(pathnames['groups'].link, pagination, filters),
                    submitRoute: getBackRoute(pathnames['groups'].link, { ...pagination, offset: 0 }, filters),
                  },
                  [pathnames['remove-group'].path]: {
                    postMethod: (ids: string[], config: Record<string, unknown>) => {
                      fetchData({ ...config, filters: { name: removingAllRows ? '' : filterValue }, orderBy });
                      removingAllRows && setFilterValue('');
                      setSelectedRows(selectedRows.filter((row) => !ids.includes(row.uuid)));
                    },
                    cancelRoute: getBackRoute(pathnames['groups'].link, pagination, filters),
                    submitRoute: getBackRoute(pathnames['groups'].link, { ...pagination, offset: 0 }, removingAllRows ? {} : filters),
                  },
                }}
              />
            </Suspense>
          </Section>
        </StackItem>
      </Stack>
    </Fragment>
  );
};

// Default export for routing
export default Groups;
