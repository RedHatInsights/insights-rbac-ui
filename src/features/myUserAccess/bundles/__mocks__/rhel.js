import React from 'react';

const InsightsBundle = () => <div id="rhel-mock">This is just a mock insights bundle module</div>;

export default InsightsBundle;
