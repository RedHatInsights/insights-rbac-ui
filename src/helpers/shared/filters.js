export const syncDefaultFiltersWithUrl = (location, navigate, keys, defaults = {}) => {
  const searchParams = new URLSearchParams(location.search);

  let filters = keys.reduce((acc, key) => {
    const values = searchParams.getAll(key);
    return {
      ...acc,
      [key]: values.length > 1 ? values : values[0],
    };
  }, {});

  Object.keys(defaults).forEach((key) => {
    const value = defaults[key];
    filters = {
      ...filters,
      [key]: Array.isArray(filters[key])
        ? [...new Set([...filters[key], ...(Array.isArray(value) ? value : [value])])]
        : (value?.length > 0 && value) || filters[key],
    };

    if (Array.isArray(value)) {
      value.forEach((item) => searchParams.getAll(key).includes(item) || searchParams.append(key, item));
    } else {
      searchParams.get(key) || (value && searchParams.set(key, value));
    }
  });

  navigate(
    {
      pathname: location.pathname,
      search: searchParams.toString(),
    },
    {
      replace: true,
    },
  );
  return filters;
};

export const areFiltersPresentInUrl = (location, keys) => {
  const searchParams = new URLSearchParams(location.search);
  return keys.some((key) => searchParams.get(key));
};

export const applyFiltersToUrl = (location, navigate, newValues) => {
  const searchParams = new URLSearchParams(location.search);
  Object.keys(newValues).forEach((key) => searchParams.delete(key));

  Object.keys(newValues).forEach((key) => {
    const value = newValues[key];

    if (Array.isArray(value)) {
      value.forEach((item) => item && searchParams.append(key, item));
    } else {
      value && searchParams.set(key, value);
    }
  });

  navigate(
    {
      pathname: location.pathname,
      search: searchParams.toString(),
    },
    {
      replace: true,
    },
  );
};
