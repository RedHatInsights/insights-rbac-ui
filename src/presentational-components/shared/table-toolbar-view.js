import React, { Fragment, useState, useEffect } from 'react';
import propTypes from 'prop-types';
import { Table, TableHeader, TableBody, TableVariant } from '@patternfly/react-table';
import { TableToolbar } from '@redhat-cloud-services/frontend-components';
import { Button, Pagination, EmptyStatePrimary } from '@patternfly/react-core';
import { ListLoader } from './loader-placeholders';
import { PlusCircleIcon } from '@patternfly/react-icons';
import { selectedRows } from '../../helpers/shared/helpers';
import Toolbar, { paginationBuilder } from './toolbar';
import EmptyWithAction from './empty-filter';
import './table-toolbar-view.scss';

export const TableToolbarView = ({
  isCompact,
  createRows,
  borders,
  columns,
  toolbarButtons,
  data,
  actionResolver,
  areActionsDisabled,
  routes,
  titlePlural,
  titleSingular,
  pagination,
  filterValue,
  isLoading,
  setFilterValue,
  checkedRows,
  isSelectable,
  fetchData,
  setCheckedItems,
  isCollapsible,
  emptyProps,
  filterPlaceholder,
  rowWrapper,
  filters,
  isFilterable,
  onShowMore,
  showMoreTitle,
  onFilter,
  onChange,
  value,
  sortBy,
  isExpandable,
  onExpand,
  hideFilterChips,
  noData,
  noDataDescription,
}) => {
  const [opened, openRow] = useState({});
  const [sortByState, setSortByState] = useState({ index: undefined, direction: undefined });
  useEffect(() => {
    setSortByState({
      ...sortBy,
      ...(sortByState.index !== undefined && sortByState),
    });
  }, [sortBy]);

  const rows = createRows(data, opened, checkedRows);

  const onCollapse = (_event, _index, isOpen, { uuid }) =>
    openRow((opened) => ({
      ...opened,
      [uuid]: isOpen,
    }));

  const renderEmpty = () => ({
    title: (
      <EmptyWithAction
        title={`No matching ${titlePlural} found`}
        description={
          noData && noDataDescription ? noDataDescription : [`This filter criteria matches no ${titlePlural}.`, `Try changing your filter settings.`]
        }
        actions={
          noData && noDataDescription
            ? undefined
            : [
                <EmptyStatePrimary key="clear-filters">
                  <Button
                    variant="link"
                    onClick={() => {
                      setFilterValue({
                        ...pagination,
                        offset: 0,
                        name: '',
                      });
                      fetchData({
                        ...pagination,
                        offset: 0,
                        name: '',
                      });
                    }}
                  >
                    Clear all filters
                  </Button>
                </EmptyStatePrimary>,
              ]
        }
      />
    ),
    props: {
      colSpan: columns.length + Boolean(onCollapse),
    },
  });

  const renderTable = () => (
    <Fragment>
      <Toolbar
        isSelectable={isSelectable}
        checkedRows={checkedRows}
        setCheckedItems={setCheckedItems}
        isLoading={isLoading || noData}
        data={data}
        titleSingular={titleSingular}
        filterValue={filterValue}
        setFilterValue={setFilterValue}
        sortBy={
          (sortByState.index !== undefined &&
            sortByState.index - isSelectable >= 0 &&
            `${sortByState.direction === 'desc' ? '-' : ''}${columns[sortByState.index - isSelectable].key}`) ||
          undefined
        }
        pagination={pagination}
        fetchData={fetchData}
        toolbarButtons={toolbarButtons}
        filterPlaceholder={filterPlaceholder}
        filters={filters}
        isFilterable={isFilterable}
        onShowMore={onShowMore}
        showMoreTitle={showMoreTitle}
        onFilter={onFilter}
        onChange={onChange}
        value={value}
        hideFilterChips={hideFilterChips}
      />
      {isLoading ? (
        <ListLoader />
      ) : (
        <Table
          canSelectAll={false}
          aria-label={`${titlePlural} table`}
          variant={isCompact ? TableVariant.compact : null}
          borders={borders}
          {...(isCollapsible && { onCollapse })}
          {...(isSelectable &&
            rows.length > 0 && {
              onSelect: (_e, isSelected, _idx, { uuid, cells: [name] }) => setCheckedItems(selectedRows([{ uuid, name }], isSelected)),
            })}
          {...(isExpandable && { onExpand })}
          rows={rows.length > 0 ? rows : [{ fullWidth: true, cells: [renderEmpty()] }]}
          cells={columns}
          {...(rows.length > 0 && { actionResolver })}
          areActionsDisabled={areActionsDisabled}
          rowWrapper={rowWrapper}
          sortBy={sortByState}
          onSort={(e, index, direction) => {
            setSortByState({ index, direction });
            fetchData({
              ...pagination,
              offset: 0,
              name: filterValue,
              orderBy: `${direction === 'desc' ? '-' : ''}${columns[index - isSelectable].key}`,
            });
          }}
        >
          <TableHeader />
          <TableBody />
        </Table>
      )}
      {!pagination.noBottom && (
        <TableToolbar>
          {!isLoading && <Pagination {...paginationBuilder(pagination, fetchData, filterValue)} variant="bottom" dropDirection="up" />}
        </TableToolbar>
      )}
    </Fragment>
  );

  return (
    <Fragment>
      {routes()}
      {!isLoading && rows.length === 0 && filterValue.length === 0 && filters.every(({ value }) => !value) ? (
        <EmptyWithAction
          title={`Configure ${titlePlural}`}
          icon={PlusCircleIcon}
          description={[`To configure user access to applications`, `create at least one ${titleSingular}`]}
          actions={toolbarButtons()[0]}
          {...emptyProps}
        />
      ) : (
        renderTable()
      )}
    </Fragment>
  );
};

TableToolbarView.propTypes = {
  ...Toolbar.propTypes,
  rowWrapper: propTypes.any,
  isCompact: propTypes.bool,
  borders: propTypes.bool,
  createRows: propTypes.func.isRequired,
  columns: propTypes.array.isRequired,
  titlePlural: propTypes.string,
  routes: propTypes.func,
  actionResolver: propTypes.func,
  areActionsDisabled: propTypes.func,
  pagination: propTypes.shape({
    noBottom: propTypes.bool,
  }),
  isExpandable: propTypes.bool,
  onExpand: propTypes.func,
  hideFilterChips: propTypes.bool,
  noDataDescription: propTypes.arrayOf(propTypes.node),
};

TableToolbarView.defaultProps = {
  ...Toolbar.defaultProps,
  isCompact: false,
  borders: true,
  routes: () => null,
  hideFilterChips: false,
};
