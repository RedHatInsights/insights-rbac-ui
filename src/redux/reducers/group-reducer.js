import {
  FETCH_GROUP,
  FETCH_SYSTEM_GROUP,
  FETCH_GROUPS,
  RESET_SELECTED_GROUP,
  FETCH_ROLES_FOR_GROUP,
  FETCH_ADD_ROLES_FOR_GROUP,
  FETCH_MEMBERS_FOR_GROUP,
} from '../../redux/action-types';
import omit from 'lodash/omit';
import { defaultSettings } from '../../helpers/shared/pagination';

// Initial State
export const groupsInitialState = {
  groups: {
    data: [],
    meta: defaultSettings,
  },
  selectedGroup: { addRoles: {}, members: { meta: defaultSettings }, pagination: defaultSettings },
  isLoading: false,
  isRecordLoading: false,
};

const setLoadingState = (state) => ({ ...state, isLoading: true });
const setRecordLoadingState = (state) => ({ ...state, isRecordLoading: true, selectedGroup: { ...state.selectedGroup, loaded: false } });
const setRecordRolesLoadingState = (state) => ({ ...state, isRecordRolesLoading: true, selectedGroup: { ...state.selectedGroup, loaded: false } });
const setGroups = (state, { payload }) => ({ ...state, groups: payload, isLoading: false });
const setSystemGroup = (state, { payload }) => ({ ...state, systemGroup: payload?.data?.filter((group) => group?.platform_default)?.[0] });
const setGroup = (state, { payload }) => ({
  ...state,
  isRecordLoading: false,
  groups: {
    ...state.groups,
    data: state.groups.data.map((group) => ({
      ...group,
      ...(payload.uuid === group.uuid && { ...payload, loaded: true }),
    })),
  },
  selectedGroup: {
    ...state.selectedGroup,
    members: { ...state.selectedGroup.members, data: payload.principals },
    ...omit(payload, ['principals', 'roles']),
    loaded: true,
    pagination: { ...state.selectedGroup.pagination, count: payload.roleCount, offset: 0 },
  },
});
const resetSelectedGroup = (state) => ({ ...state, selectedGroup: undefined });
const setRolesForGroup = (state, { payload }) => ({
  ...state,
  isRecordRolesLoading: false,
  selectedGroup: { ...state.selectedGroup, roles: payload.data, pagination: payload.meta, loaded: true },
});

const setMembersForGroupLoading = (state = {}) => ({
  ...state,
  selectedGroup: {
    ...(state.selectedGroup || {}),
    members: { isLoading: true },
  },
});
const setMembersForGroup = (state, { payload }) => ({
  ...state,
  selectedGroup: {
    ...(state.selectedGroup || {}),
    members: {
      isLoading: false,
      ...payload,
    },
  },
});

const setAddRolesLoading = (state) => ({
  ...state,
  selectedGroup: { ...state.selectedGroup, addRoles: { loaded: false } },
});
const setAddRolesForGroup = (state, { payload }) => ({
  ...state,
  selectedGroup: { ...state.selectedGroup, addRoles: { roles: payload.data, pagination: payload.meta, loaded: true } },
});

export default {
  [`${FETCH_GROUPS}_PENDING`]: setLoadingState,
  [`${FETCH_GROUPS}_FULFILLED`]: setGroups,
  [`${FETCH_SYSTEM_GROUP}_PENDING`]: setLoadingState,
  [`${FETCH_SYSTEM_GROUP}_FULFILLED`]: setSystemGroup,
  [`${FETCH_GROUP}_PENDING`]: setRecordLoadingState,
  [`${FETCH_GROUP}_FULFILLED`]: setGroup,
  [`${FETCH_ROLES_FOR_GROUP}_PENDING`]: setRecordRolesLoadingState,
  [`${FETCH_ROLES_FOR_GROUP}_FULFILLED`]: setRolesForGroup,
  [`${FETCH_MEMBERS_FOR_GROUP}_PENDING`]: setMembersForGroupLoading,
  [`${FETCH_MEMBERS_FOR_GROUP}_FULFILLED`]: setMembersForGroup,
  [`${FETCH_ADD_ROLES_FOR_GROUP}_PENDING`]: setAddRolesLoading,
  [`${FETCH_ADD_ROLES_FOR_GROUP}_FULFILLED`]: setAddRolesForGroup,
  [RESET_SELECTED_GROUP]: resetSelectedGroup,
};
