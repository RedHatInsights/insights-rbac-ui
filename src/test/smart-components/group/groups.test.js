import React from 'react';
import { act } from 'react-dom/test-utils';
import thunk from 'redux-thunk';
import { MemoryRouter as Router } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import promiseMiddleware from 'redux-promise-middleware';
import Groups from '../../../smart-components/group/groups';
import { mock } from '../../__mocks__/apiMock';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications';
import { groupsInitialState } from '../../../redux/reducers/group-reducer';
import { TableToolbarView } from '../../../presentational-components/shared/table-toolbar-view';
import { getUserMock } from '../../../../config/setupTests';

describe('<Groups />', () => {
  let enhanceState;
  const middlewares = [thunk, promiseMiddleware, notificationsMiddleware()];
  let mockStore;
  let initialState;

  beforeEach(() => {
    enhanceState = {
      data: [
        {
          uuid: '19eccddf-7e6a-41a2-b050-3e67f0af4ed1',
          name: 'Custom default user access',
          description: 'Default group that contains default permissions for new users.',
          principalCount: 0,
          platform_default: true,
          roleCount: 6,
          created: '2019-12-03T18:56:01.184350Z',
          modified: '2020-02-06T08:50:12.064832Z',
          system: false,
        },
      ],
      meta: {
        count: 153,
        limit: 10,
        offset: 0,
      },
    };
    mockStore = configureStore(middlewares);
    initialState = { groupReducer: { ...groupsInitialState, groups: enhanceState, systemGroup: enhanceState.data[0] } };
  });

  afterEach(() => {
    mock.reset();
  });

  it('should render group list correctly', async () => {
    mock.onGet(`/api/rbac/v1/groups/?limit=10&offset=0&name=`).replyOnce(200, {});
    mock.onGet(`/api/rbac/v1/groups/?limit=1&name=default&name_match=partial`).replyOnce(200, {});
    let wrapper;
    const store = mockStore(initialState);
    await act(async () => {
      wrapper = mount(
        <Provider store={store}>
          <Router initialEntries={['/groups']}>
            <Groups />
          </Router>
        </Provider>
      );
    });
    expect(wrapper.find(TableToolbarView)).toHaveLength(1);
  });

  it('should fetch groups on mount', () => {
    const store = mockStore(initialState);
    mock.onGet(`/api/rbac/v1/groups/?limit=10&offset=0&name=`).replyOnce(200, {});
    mock.onGet(`/api/rbac/v1/groups/?limit=1&name=default&name_match=partial`).replyOnce(200, {});
    mount(
      <Provider store={store}>
        <Router>
          <Groups />
        </Router>
        ,
      </Provider>
    );
    const expectedPayload = [{ type: 'FETCH_GROUPS_PENDING' }, { type: 'FETCH_SYSTEM_GROUP_PENDING' }];
    expect(store.getActions()).toEqual(expectedPayload);
  });

  it('should fetch groups on next page click', async () => {
    const store = mockStore(initialState);
    mock.onGet(`/api/rbac/v1/groups/?limit=10&offset=0&name=`).replyOnce(200, {});
    mock.onGet(`/api/rbac/v1/groups/?limit=1&name=default&name_match=partial`).replyOnce(200, {});
    mock.onGet(`/api/rbac/v1/groups/?limit=10&offset=10&name=&order_by=`).replyOnce(200, {});
    let wrapper;
    await act(async () => {
      wrapper = mount(
        <Provider store={store}>
          <Router initialEntries={['/groups']}>
            <Groups />
          </Router>
        </Provider>
      );
    });
    store.clearActions();
    await act(async () => {
      wrapper.find('.pf-c-pagination__nav .pf-c-button').at(1).simulate('click');
    });
    const expectedPayloadAfter = [
      { type: 'FETCH_GROUPS_PENDING' },
      {
        type: 'FETCH_GROUPS_FULFILLED',
        payload: {
          ...getUserMock,
        },
      },
    ];
    expect(store.getActions()).toEqual(expectedPayloadAfter);
  });

  it('should fetch groups on filter and cancel filter', async () => {
    jest.useFakeTimers();
    const store = mockStore(initialState);
    const filterValue = 'filterValue';
    mock.onGet(`/api/rbac/v1/groups/?limit=10&offset=0&name=`).replyOnce(200, {});
    mock.onGet(`/api/rbac/v1/groups/?limit=1&name=default&name_match=partial`).replyOnce(200, {});
    mock.onGet(`/api/rbac/v1/groups/?limit=10&offset=0&name=${filterValue}`).replyOnce(200, {});
    mock.onGet(`/api/rbac/v1/groups/?limit=10&offset=0&name=`).replyOnce(200, {});
    let wrapper;
    await act(async () => {
      wrapper = mount(
        <Provider store={store}>
          <Router initialEntries={['/groups']}>
            <Groups />
          </Router>
        </Provider>
      );
    });
    store.clearActions();
    await act(async () => {
      wrapper.find('input#filter-by-string').simulate('change', { target: { value: filterValue } });
    });
    const expectedFilterPayload = [{ type: 'FETCH_GROUPS_PENDING' }];
    jest.runAllTimers();
    expect(store.getActions()).toEqual(expectedFilterPayload);
    store.clearActions();
    wrapper.update();
    await act(async () => {
      wrapper.find('#ins-primary-data-toolbar .ins-c-chip-filters').simulate('click');
    });
    const expectedCancelPayload = [
      {
        type: 'FETCH_GROUPS_FULFILLED',
        payload: {
          ...getUserMock,
        },
      },
    ];
    expect(store.getActions()).toEqual(expectedCancelPayload);
  });

  it('should fetch groups on sort click', async () => {
    const store = mockStore(initialState);
    mock.onGet(`/api/rbac/v1/groups/?limit=10&offset=0&name=`).replyOnce(200, {});
    mock.onGet(`/api/rbac/v1/groups/?limit=1&name=default&name_match=partial`).replyOnce(200, {});
    mock.onGet(`/api/rbac/v1/groups/?limit=10&offset=0&name=&order_by=name`).replyOnce(200, {});
    let wrapper;
    await act(async () => {
      wrapper = mount(
        <Provider store={store}>
          <Router initialEntries={['/groups']}>
            <Groups />
          </Router>
        </Provider>
      );
    });
    store.clearActions();
    await act(async () => {
      wrapper.find('span.pf-c-table__sort-indicator').first().simulate('click');
    });
    const expectedPayloadAfter = [
      { type: 'FETCH_GROUPS_PENDING' },
      {
        type: 'FETCH_GROUPS_FULFILLED',
        payload: {
          ...getUserMock,
        },
      },
    ];
    expect(store.getActions()).toEqual(expectedPayloadAfter);
  });
});
