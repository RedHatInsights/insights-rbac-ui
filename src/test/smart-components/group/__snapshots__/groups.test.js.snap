// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Groups /> should render correctly 1`] = `
<Groups
  fetchGroups={[Function]}
  groups={Array []}
  isLoading={false}
  pagination={
    Object {
      "count": 0,
      "limit": 10,
      "offset": 0,
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  userIdentity={Object {}}
/>
`;

exports[`<Groups /> should render correctly in loading state 1`] = `
<Groups
  fetchGroups={[Function]}
  groups={Array []}
  isLoading={false}
  pagination={
    Object {
      "count": 0,
      "limit": 10,
      "offset": 0,
    }
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  userIdentity={Object {}}
/>
`;
